exercise/api/views.py

```diff
@@ -222,6 +226,77 @@ def create(self, request, exercise_id, version):
 
             return Response(status=status.HTTP_201_CREATED)
 
+    @action(
+        detail=False,
+        url_path='submit',
+        url_name='submit',
+        methods=['post'],
+    )
+    def submit(self, request, *args, **kwargs):
+        """Submit a new solution to the exercise for grading.
+        Students are allowed to submit via this endpoint. In a group submission,
+        the student group ID is defined with the POST parameter _aplus_group.
+        The POST data is used as the submission data and files may be uploaded too.
+        """
+        # Stop submit trials for e.g. chapters.
+        # However, allow posts from exercises switched to maintenance status.
+        if not self.exercise.is_submittable:
+            return self.http_method_not_allowed(request, *args, **kwargs)
+
+        data = None
+        status_code = None
+        headers = None
+        submission_status, issues, students = (
+            self.exercise.check_submission_allowed(request.user.userprofile, request)
+        )
+        if submission_status == self.exercise.SUBMIT_STATUS.ALLOWED:
+            new_submission = Submission.objects.create_from_post(
+                self.exercise, students, request)
+            if new_submission:
+                page = self.exercise.grade(request, new_submission, use_messages=False)
+
+                # Enroll after succesfully submitting to the enrollment exercise.
+                if self.exercise.status in (
+                    LearningObject.STATUS.ENROLLMENT,
+                    LearningObject.STATUS.ENROLLMENT_EXTERNAL,
+                ) and new_submission.status == Submission.STATUS.READY:
+                    self.instance.enroll_student(request.user)
+
+                status_code = status.HTTP_201_CREATED
+                headers = {
+                    'Location': reverse('api:submission-detail', kwargs={
+                                        'submission_id': new_submission.id,
+                                        }, request=request),
+                }
+                if page.errors:
+                    data = {'errors': page.errors}
+            else:
+                data = {
+                    'detail': _("The submission could not be saved for some reason. "
+                                "The submission was not registered."),
+                }
+                status_code = status.HTTP_500_INTERNAL_SERVER_ERROR
```

Comment:
`Submission.objects.create_from_post` does not validate POST data. The grader may reject submissions at the start of grading if the submission is missing required fields.

It could be better if A+ could do basic validation for submissions before sending them to the grader, but it doesn't currently work like that.

I was wondering if I should use HTTP 400 response for the case when students aren't allowed to submit due to deadlines or max submissions limit. It is not really an error in the HTTP request like the 400 code suggests, so is "200 OK" better? The error messages for the user are included in the response body anyway.

Reply:
400 should be fine when the client is not supposed to repeat the same post as is. https://stackoverflow.com/a/8475236 and https://stackoverflow.com/a/6123801